File: prog1/TrabFinal/Pilha.c
/* Define as funções declaradas em pilha.h */
#include <stdlib.h>
#include <stdio.h>
#include "Pilha.h"
#include <stdbool.h>



/*Recebe inteiro, aloca memoria para um itemP, atribui o valor e retorna o itemP*/
struct candidata *criaCandidata(int valor) {
        struct candidata *novoCandidata;

        if (!(novoCandidata = malloc(sizeof(struct candidata)))) { /*aloca espaco na memoria para o itemP a ser adicionado*/
                return NULL;
        }

        novoCandidata->id = valor; /*criacao no valor passado pelo usuario dentro da pilha*/
        novoCandidata->proximo = NULL; /*o proximo eh 0 para ter controle*/

        return novoCandidata;
}

/*Aloca memoria para uma pilha, zera o tamanho e atribui o ponteiros topo para NULO*/
struct pilha *inicializaPilha() {
        struct pilha *p;

        if (!(p = malloc(sizeof(struct pilha)))) { /*aloca espaco na memoria para a pilha*/
                return NULL;
        }

        p->tamanho = 0; /*tamanho inicializado com 0*/
        p->topo = NULL; /*topo sendo 0 para controle*/

        return p;
}

/*Retorna true para pilha vazia e false para pilha com ao menos um itemP*/
bool pilhaVazia(struct pilha *p) {
        if (p->topo == 0) {
                return true; /*pilha vazia*/
        }
        return false; /*pilha com ao menos um itemP*/
}


/*Dada uma pilha P, insere itemP com valor x chamando criaItem no topo da pilha;*/ 
/*retorna 1 se x foi inserido com sucesso, 0 se x nao foi inserido.*/
int push(struct pilha *p, int x) {
        struct candidata *novo = criaCandidata(x);

        if (novo == NULL) { /*falha na alocacao de memoria*/
                return 0;
        }

        if (p->tamanho == 0) { /*insere o primeiro elemento no topo*/

                p->topo = novo;
                p->topo->proximo = NULL;
                p->tamanho++;

                return 1; /*empilhado com sucesso*/

        } else { /*se ja tiver um elemento*/

                novo->proximo = p->topo;
                p->topo = novo;
                p->tamanho++;

                return 1; /*empilhado com sucesso*/
        }

        return 0; /*pilha vazia, nao eh possivel enfileirar*/
}

/*Dada uma pilha P, remove elemento do topo, caso a pilha nao esteja vazia;*/ 
/* retorna 1 se o elemento foi removido ou 0 se o elemento nao foi removido.*/
/* O ponteiro para inteiro 'dado' retorna o valor do topo...*/
int pop(struct pilha *p, int *dado) {
        struct candidata *dadoRemov; /*criacao do itemP de remocao*/

        /*verifica pilha vazia e se falso realiza as operacoes de remover da pilha*/
        if (pilhaVazia(p) == false) {

                dadoRemov = p->topo; /*comecar a remover do topo*/

                *dado = dadoRemov->valor; 
                p->topo = dadoRemov->proximo;

                free(dadoRemov); /*liberacao da memoria alocada*/

                p->tamanho--;

                return 1;
        }
        return 0;
}

void imprimePilha(struct pilha *p) {
        struct candidata *aux;
        aux = p->topo;

        printf("Impressão dos elementos presentes na pilha: \n");

        while(aux) {
                printf("Elemento da pilha: %d\n", aux->id);

                aux = aux->proximo;
        }
}

/*Imprime o topo da pilha*/
void imprimeTopo(struct pilha *p) {
        if ((p != NULL) && (p->topo->valor != 0)) {
                printf("Topo da pilha: %d\n", p->topo->id);
        } else {
                printf("Nao foi incializado ou está vazio\n");
        }
}

/*Retorna o tamanho da pilha*/
int tamanhoPilha(struct pilha *p) {
        return (p->tamanho);
}

/*Destroi cada itemP da pilha (se nao vazia) e destroi a pilha*/
void destroiPilha(struct pilha *p) {
        /*criacao de uma nova pilha que eh a copia da que utilizamos para realizar as outras funcoes*/
        /*alocar memoria para a pilha nova*/
        struct candidata *dadoRemov;

        /*verifica pilha vazia e se falso realiza as operacoes de infileirar na fila*/
        if (pilhaVazia(p) == false) {
                /*destruir cada itemP*/
                /*fazer um while que vai alocando um elemento por vez ate todos estarem NULL*/
                while (p->tamanho != 0) {
                        dadoRemov = p->topo; 
                        p->topo = p->topo->proximo;

                        free(dadoRemov); /*libera memoria que foi alocada do elemento*/

                        p->tamanho--;
                }

                printf("Pilha destruida\n");

        } else {

                printf("Pilha vazia\n");
        }

        free(p); /*libera memoria que foi alocada do elemento*/
}

File: paradigmas/estruturas/Pilha.java
class Nodo {
    String elemento;
    Nodo proximo = null;

    public Nodo(String elemento) {
        this.elemento = elemento;
    }
}

public class Pilha {
    private Nodo topo = null;
    private String nome = "lista";

    // Método para empilhar um elemento na pilha
    public void push(String elemento) {
        Nodo no = new Nodo(elemento);
        no.proximo = this.topo;
        this.topo = no;
    }

    // Método para desempilhar o elemento do topo da pilha
    public void pop() {
        if (topo != null) {
            Nodo aux = this.topo;
            this.topo = aux.proximo;
            aux = null; // Remove a referência para ajudar o Garbage Collector
        } else {
            System.out.println("A pilha está vazia.");
        }
    }

    // Método para retornar o elemento do topo sem remover
    public String top() {
        if (topo != null) {
            return topo.elemento;
        } else {
            System.out.println("A pilha está vazia.");
            return null;
        }
    }

    // Método para verificar se a pilha está vazia
    public boolean empty() {
        return topo == null;
    }

    // Método para retornar o tamanho da pilha
    public int size() {
        int contador = 0;
        Nodo atual = topo;
        
        while (atual != null) {
            contador++;
            atual = atual.proximo;
        }
        
        return contador;
    }

    // Método para imprimir todos os elementos da pilha
    public void imprimePilha() {
        if (empty()) {
            System.out.println("A pilha está vazia.");
            return;
        }

        Nodo atual = topo;
        System.out.println("Elementos da pilha:");
        while (atual != null) {
            System.out.println(atual.elemento);
            atual = atual.proximo;
        }
    }

    // Método main para testar a pilha
    public static void main(String[] args) {
        Pilha pilha = new Pilha();
        Pilha p = new Pilha();

        for(int i = 0; i < 10; i++) {
            p.push("Elemento " + i);
        }

        pilha.push("Elemento 1");
        pilha.push("Elemento 2");
        pilha.push("Elemento 3");
        pilha.push("Elemento 4");
        pilha.push("Elemento 5");
        pilha.push("Elemento 6");
        pilha.push("Elemento 7");
        pilha.push("Elemento 8");
        pilha.push("Elemento 9");
        pilha.push("Elemento 10");

        while (!pilha.empty()) {
            pilha.pop();
        }

        System.out.println("Elemento do topo: " + pilha.top()); // Saída: Elemento 3
        System.out.println("Tamanho da pilha: " + pilha.size()); // Saída: 3

        pilha.imprimePilha();

        pilha.pop();
        System.out.println("\nDepois de um pop:");
        pilha.imprimePilha();
        p.imprimePilha();

        System.out.println("A pilha está vazia? " + pilha.empty()); // Saída: false
    }
}
